metautil/type_conversions.go

```diff
@@ -0,0 +1,67 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package metautil
+
+import (
+	"fmt"
+	"reflect"
+)
+
+func convertValue(value reflect.Value, outputType reflect.Type) (reflect.Value, error) {
+	inputType := value.Type()
+	if inputType == outputType {
+		return value, nil
+	}
+
+	var nullValue reflect.Value
+	switch value.Kind() {
+	case reflect.Slice, reflect.Array:
+		if outputType.Kind() != reflect.Array && outputType.Kind() != reflect.Slice {
+			break
+		}
+		outputValue := reflect.MakeSlice(outputType, 0, value.Len())
+		for i := 0; i < value.Len(); i++ {
+			convertedElem, err := convertValue(value.Index(i), outputType.Elem())
+			if err != nil {
+				return nullValue, err
+			}
+			outputValue = reflect.Append(outputValue, convertedElem)
+		}
+		return outputValue, nil
+	case reflect.Interface:
+		return convertValue(value.Elem(), outputType)
+	}
+	return nullValue, fmt.Errorf(`cannot convert value "%v" into a %v`, value, outputType)
+}
+
+func SetValueFromAttribute(snapName string, ifaceName string, attrName string, attrVal interface{}, val interface{}) error {
+	rt := reflect.TypeOf(val)
+	if rt.Kind() != reflect.Ptr || val == nil {
+		return fmt.Errorf("internal error: cannot get %q attribute of interface %q with non-pointer value", attrName, ifaceName)
```

Comment:
I think this error should be made more generic since it's moved to metautil now? snapName, ifaceName, attrName should be dropped from the arguments?

Reply:
I don't think so, because this method is very specific to snap interfaces. Unless I misunderstood @pedronis's suggestion above. :-)

Reply:
attribute in the in the name indeed refers to interface attributes specifically, so I think until we have another use case the approach is correct. OTOH this does need a doc comment